src/core/primitives.rs

```diff
@@ -0,0 +1,612 @@
+        symmetric_crypto::aes_256_gcm_pure::Aes256GcmCrypto, CsRng,
+    };
+
+    //
+    // Define types and constants used in the tests
+    //
+    const TAG_LENGTH: usize = 32;
+    const SYM_KEY_LENGTH: usize = 32;
+    type KeyPair = X25519KeyPair;
+    #[allow(clippy::upper_case_acronyms)]
+    type DEM = Aes256GcmCrypto;
+
+    #[test]
+    fn test_kyber() {
+        let mut rng = CsRng::from_entropy();
+        let keypair = pqc_kyber::keypair(&mut rng);
+        let (ct, ss) = pqc_kyber::encapsulate(&keypair.public, &mut rng).unwrap();
+        let res = pqc_kyber::decapsulate(&ct, &keypair.secret).unwrap();
+        assert_eq!(ss, res, "Decapsulation failed!");
+    }
+
+    #[test]
+    fn test_cover_crypt() -> Result<(), Error> {
+        let admin_partition = Partition(b"admin".to_vec());
+        let dev_partition = Partition(b"dev".to_vec());
+        // partition list
+        let partitions_set = HashMap::from([
+            (admin_partition.clone(), EncryptionHint::Hybridized),
+            (dev_partition.clone(), EncryptionHint::Classic),
+        ]);
+        // user list
+        let users_set = vec![
+            HashSet::from([dev_partition.clone()]),
+            HashSet::from([admin_partition.clone(), dev_partition.clone()]),
+        ];
+        // target set
+        let admin_target_set = HashSet::from([admin_partition.clone()]);
+        // secure random number generator
+        let mut rng = CsRng::from_entropy();
+        // setup scheme
+        let (mut msk, mut mpk) = setup!(&mut rng, &partitions_set);
+        let admin_secret_subkeys = msk.x.get(&admin_partition);
+        assert!(admin_secret_subkeys.is_some());
+        assert!(admin_secret_subkeys.unwrap().0.is_some());
+        let dev_secret_subkeys = msk.x.get(&dev_partition);
+        assert!(dev_secret_subkeys.is_some());
+        assert!(dev_secret_subkeys.unwrap().0.is_none());
+        // generate user secret keys
+        let mut usk0 = keygen!(&mut rng, &msk, &users_set[0]);
+        let usk1 = keygen!(&mut rng, &msk, &users_set[1]);
+        // encapsulate for the target set
+        let (sym_key, encapsulation) = encaps!(&mut rng, &mpk, &admin_target_set);
+
+        assert_eq!(encapsulation.E.len(), 1);
+        for key_encapsulation in &encapsulation.E {
+            if let KeyEncapsulation::ClassicEncapsulation(_) = key_encapsulation {
+                panic!("Wrong hybridization type");
+            }
+        }
+
+        // decapsulate for users 1 and 3
+        let res0 = decaps!(&usk0, &encapsulation);
+
+        assert!(res0.is_err(), "User 0 shouldn't be able to decapsulate!");
+
+        let res1 = decaps!(&usk1, &encapsulation)?;
+
+        assert_eq!(sym_key, res1, "Wrong decapsulation for user 1!");
+
+        // Change partition
+        let client_partition = Partition(b"client".to_vec());
+        let new_partitions_set = HashMap::from([
+            (dev_partition, EncryptionHint::Classic),
```

Comment:
Could you try upgrading a partition from classic to hybrid / hybrid to classic in a rotate? Not sure what is suppose to be the rule here but we should have a test to check that it's not something wrong?

Reply:
Sure, i thought i did something like that but i might have slipped  throught